Feat lineType catalog

EstherDarkish opened the pull request.

archendel commented on src/components/dialogs/line-type-catalog/line-type-catalog-selector-dialog.js:
```diff
@@ -0,0 +1,275 @@
+            disabledSave={!selectedRow}
+        >
+            <div style={{ height: '100%' }}>
+                <CustomAGGrid
+                    ref={gridRef}
+                    rowData={
+                        tabIndex ===
+                        LineTypeCatalogSelectorDialogTabs.AERIAL_TAB
+                            ? rowDataAerialTab
+                            : rowDataUndergroundTab
+                    }
+                    defaultColDef={defaultColDef}
+                    columnDefs={columns}
+                    rowSelection="single"
+                    onSelectionChanged={onSelectionChanged}
+                    onGridReady={highlightSelectedRow} // Highlights the preselected row when AGGrid is ready
+                />
```
we might gain some performance doing this : 

```
const Table = useCallback((tabIndex) => {
   const data =  tabIndex ===
                        LineTypeCatalogSelectorDialogTabs.AERIAL_TAB
                            ? rowDataAerialTab
                            : rowDataUndergroundTab;

    return (
        <CustomAGGrid
                    ref={gridRef}
                    rowData={data}
                    defaultColDef={defaultColDef}
                    columnDefs={columns}
                    rowSelection="single"
                    onSelectionChanged={onSelectionChanged}
                    onGridReady={highlightSelectedRow} // Highlights the preselected row when AGGrid is ready
                />
    )
} [allDependencies]);
```

archendel commented on src/components/dialogs/line-type-catalog/line-type-segment-creation.js:
```diff
@@ -0,0 +1,109 @@
+}) => {
+    const classes = useStyles();
+
+    const segmentDistanceField = (
```
I'm confused with this PR, sometimes you create const to store inputs, sometimes you create them directly in the render block

I think we should stick to one way to do it

archendel replied:
same thing in "line-type-segment-form.js", it would greatly reduce the JSX complexity

EstherDarkish replied:
Done.

archendel commented on src/components/dialogs/line-type-catalog/line-type-segment-creation.js:
```diff
@@ -0,0 +1,109 @@
+        [index, onEditButtonClick]
+    );
+
+    const watchDistance = useWatch({
```
create watchers at the start of the component so the developpers can easily see which data will cause the component to rerender

archendel reviewed (changes requested):
When opening catalog, if catalog can't be fetched, no error message appears.
Then when you try to select a line type, the tables are empty. It can be confusing

EstherDarkish commented:
> When opening catalog, if catalog can't be fetched, no error message appears. Then when you try to select a line type, the tables are empty. It can be confusing

Fixed. I added a snackbar error when the catalog fetch fails.

EstherDarkish commented on src/components/utils/utils.js:
```diff
@@ -98,3 +98,46 @@ export const buildNewBusbarSections = (
     return newBusbarSections;
 };
+
+export function calculateResistance(distance, linearResistance) {
+    if (
+        distance === undefined ||
+        isNaN(distance) ||
+        linearResistance === undefined ||
+        isNaN(linearResistance)
+    ) {
+        return 0;
+    }
+    return Number(distance) * Number(linearResistance);
+}
+
+export function calculateReactance(distance, linearReactance) {
+    if (
+        distance === undefined ||
+        isNaN(distance) ||
+        linearReactance === undefined ||
+        isNaN(linearReactance)
+    ) {
+        return 0;
+    }
+    return Number(distance) * Number(linearReactance);
+}
```
They do the same thing, but they have different roles. I am not a fan of either of those options :

- Use only one function called "calculateReactanceOrResistance" => What's the next step ? "calculateReactanceOrResistanceOrRadiusOrWhateverThatsJustMathAnyway" ? :)
- Do not use a function and directly do the math => With specific functions, we can use and maintain them easily.

Do you have another better solution ?

archendel replied:
I am wondering if this wouldn't do the same thing, and would make it a one line operation instead of duplicate functions :
 
distance * linearReactance || 0

if one of them is not a number or is undefined, it will return NaN, and NaN || 0 = 0

EstherDarkish commented on src/utils/rest-api.js:
```diff
@@ -159,9 +159,8 @@ export function fetchValidateUser(user) {
-            } else {
-                throw error;
             }
+            throw error;
```
That's totally painless and it is something that was forgotten in a previous PR. Do you want me to create a new PR only for this change ?

EstherDarkish replied:
Ok, rollbacked.

EstherDarkish commented on src/components/dialogs/commons/basicModificationDialog.js:
```diff
@@ -0,0 +1,125 @@
+ * @param {Array} dialogProps props that are forwarded to the MUI Dialog component
+ * @param {Boolean} isDataFetching props to display loading
+ */
+const BasicModificationDialog = ({
```
I tried to refactor these and include one in the other, but could not do it in time. This should be done in a separate technical story.

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-segment-dialog.js:
```diff
@@ -0,0 +1,104 @@
+};
+
+const formSchema = yup.object().shape({
+    [TOTAL_RESISTANCE]: yup.number().required(),
+    [TOTAL_REACTANCE]: yup.number().required(),
+    [TOTAL_SUSCEPTANCE]: yup.number().required(),
```
These fields are read-only and cannot have en empty string in them.

EstherDarkish commented on src/components/utils/rhf-inputs/read-only/read-only-input.js:
```diff
@@ -0,0 +1,54 @@
+const useStyles = makeStyles((theme) => ({
+    readOnlyTextField: {
+        '& fieldset': {
+            padding: '0 !important',
```
It is not, it is old code that I used to test changes. I removed it.

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-segment-form.js:
```diff
@@ -0,0 +1,300 @@
+            return matchingCatalogEntry?.linearCapacity ?? 0;
+        },
+        [lineTypeCatalog]
+    );
```
I refactored these. Thank you

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-segment-form.js:
```diff
@@ -0,0 +1,300 @@
+
+    const onSelectCatalogLine = useCallback(
+        (selectedLine) => {
+            if (selectedLine) {
```
We cannot select no line.I'll keep the "if (selectedLine)" to prevent errors in the futur if there's some bad usages of these functions.

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-catalog-selector-dialog.js:
```diff
@@ -0,0 +1,275 @@
+    // to the selected line.
+    const highlightSelectedRow = useCallback(() => {
+        const rowToHightlight = selectedRow ?? props?.preselectedRow;
+        if (rowToHightlight && props?.rowData) {
+            const currentTabKind =
+                tabIndex === LineTypeCatalogSelectorDialogTabs.AERIAL_TAB
+                    ? 'AERIAL'
+                    : 'UNDERGROUND';
+            if (rowToHightlight.kind === currentTabKind) {
```
The first tests if we have a selected row, the second tests if we are in the correct tab.

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-catalog-selector-dialog.js:
```diff
@@ -0,0 +1,275 @@
+            sx={{
+                display: 'flex',
+                flexWrap: 'wrap',
+                gap: '15px',
```
That's some code lifted from other places. Because 15 is not cleanly divisible by 4, I'd rather left this like this and we could clean all of these instances in another PR.

EstherDarkish commented on src/components/dialogs/line-type-catalog/line-type-catalog-selector-dialog.js:
```diff
@@ -0,0 +1,275 @@
+        selectedRow,
+        tabIndex,
+        props.preselectedRow,
+        props.rowData,
```
Yes, cleaned it.

sBouzols commented on src/components/dialogs/line-type-catalog/line-type-catalog-selector-dialog.js:
```diff
@@ -0,0 +1,275 @@
+
+const LineTypeCatalogSelectorDialog = (props) => {
+    const intl = useIntl();
+    const { onClose, onSelectLine } = props;
```
fixed some warning because of this and dialogProps forwarding

archendel commented on src/components/utils/rhf-inputs/read-only/read-only-input.js:
```diff
@@ -0,0 +1,54 @@
+                readOnly: true,
+                disableUnderline: true,
+            }}
+            multiline
```
we almost don't do it anywhere else in the app, should we consider using OverflowableText instead ?

sBouzols reviewed (approved):
Code Review OK
Tests OK
Console warning check OK
